experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts

```diff
@@ -0,0 +1,476 @@
+					"subtreeChanging",
+					(anchorNode: AnchorNode) => listener(anchorNode),
+				);
+				return unsubscribeFromSubtreeChange;
+			}
+			default:
+				unreachableCase(eventName);
+		}
+	}
+}
+
+export class LazyMap<TSchema extends MapSchema>
+	extends LazyTree<TSchema>
+	implements MapNode<TSchema>
+{
+	public constructor(
+		context: Context,
+		schema: TSchema,
+		cursor: ITreeSubscriptionCursor,
+		anchorNode: AnchorNode,
+		anchor: Anchor,
+	) {
+		super(context, schema, cursor, anchorNode, anchor);
+
+		// Setup JS Object API:
+		makePropertyEnumerableOwn(this, "asObject", LazyMap.prototype);
+	}
+
+	public get(key: FieldKey): TypedField<TSchema["mapFields"]> {
+		return inCursorField(this[cursorSymbol], key, (cursor) =>
+			makeField(this.context, this.schema.mapFields, cursor),
+		) as TypedField<TSchema["mapFields"]>;
+	}
+
+	// TODO: when appropriate add setter that delegates to field kind specific setter.
+	// public set(key: FieldKey, content: FlexibleFieldContent<TSchema["mapFields"]>): void {
+	// 	const field = this.get(key);
+	// 	if (field.is(SchemaBuilder.fieldOptional(...this.schema.mapFields.allowedTypes))) {
+	// 		field.setContent(content);
+	// 	} else {
+	// 		assert(
+	// 			field.is(SchemaBuilder.fieldSequence(...this.schema.mapFields.allowedTypes)),
+	// 			"unexpected map field kind",
+	// 		);
+	// 		// TODO: fix merge semantics.
+	// 		field.replaceRange(0, field.length, content as Iterable<ContextuallyTypedNodeData>);
+	// 	}
+	// }
+
+	public [Symbol.iterator](): IterableIterator<TypedField<TSchema["mapFields"]>> {
+		return super[Symbol.iterator]() as IterableIterator<TypedField<TSchema["mapFields"]>>;
+	}
+
+	public get asObject(): {
+		readonly [P in FieldKey]?: UnboxField<TSchema["mapFields"]>;
+	} {
+		const record: Record<FieldKey, UnboxField<TSchema["mapFields"]> | undefined> =
+			Object.create(null);
+
+		forEachField(this[cursorSymbol], (cursor) => {
+			Object.defineProperty(record, cursor.getFieldKey(), {
+				value: unboxedField(this.context, this.schema.mapFields, cursor),
+				configurable: true,
+				enumerable: true,
+			});
+		});
+		return record;
+	}
+}
+
+export class LazyLeaf<TSchema extends LeafSchema>
+	extends LazyTree<TSchema>
+	implements Leaf<TSchema>
+{
+	public constructor(
+		context: Context,
+		schema: TSchema,
+		cursor: ITreeSubscriptionCursor,
+		anchorNode: AnchorNode,
+		anchor: Anchor,
+	) {
+		super(context, schema, cursor, anchorNode, anchor);
+
+		// Setup JS Object API:
+		makePropertyEnumerableOwn(this, "value", LazyTree.prototype);
+	}
+
+	public override get value(): SchemaAware.InternalTypes.TypedValue<TSchema["leafValue"]> {
+		return super.value as SchemaAware.InternalTypes.TypedValue<TSchema["leafValue"]>;
+	}
+}
+
+export class LazyFieldNode<TSchema extends FieldNodeSchema>
+	extends LazyTree<TSchema>
+	implements FieldNode<TSchema>
+{
+	public get content(): UnboxField<TSchema["structFieldsObject"][""]> {
```

Comment:
Do that and you have to package export that helper to make API extractor happy, document it, check how it impacts intelisense, give it a name that makes sense in a much larger scope, then delete it once we refactor TreeSchema to split it up properly. I don't think its worth it. There is already a bunch of notes and todos about this on FieldNodeSchema, which is the type being worked on here.

We do have "EmptyKey" to provide self documenting use of this pattern in other context: it just doesn't work here due to annoyances with typescript const string type handling and branding.

I have added even more doc comments and TODOs so code like this can get intellisense on all the parts of this expression, each explaining the situation.

Reply:
Sounds good, thanks!